### ISSUE TITLE
Add timestamps to `*XELB-DEBUG*` 

### ISSUE THREAD

**Issue by emacksnotes** [ch11ng/exwm#610]
### Add timestamps to `*XELB-DEBUG*` 

While looking at `*XELB-DEBUG*`, I was a bit clueless about how the logs are to be read.  I really didn't know if the logs are in chronological order or reverse chronological order.  

Consider adding a timestamp to debug messages.  If that is heavyweight, consider adding a buffer header line that informs the user  the order in which the log buffer has to be read.

**Comment by medranocalvo** [comment]
My gut feeling is that timestamps would be too heavyweight, both for processing and visually.  The informational header line idea is superb.

Please, try the following in `*scratch*`:

```elisp
(defun exwm-debug--hide-header-line (button)
  (setq-local header-line-format nil)
  (force-mode-line-update))

(setq-local header-line-format 
            (with-temp-buffer
              (insert "Messages appear in chronological order. ")
              (insert-text-button "Hide hint"
                                  'face custom-button
                                  'follow-link t
                                  'action 'exwm-debug--hide-header-line)
              (buffer-string)))
```

Could you suggest a better informational message?

**Comment by emacksnotes** [comment]
> My gut feeling is that timestamps would be too heavyweight, both for processing and visually. The informational header line idea is superb.

I have a scenario where the timestamp will be very useful.  

I have run in to 

> "'XXX is not responding.  Would you like to kill it? "  (from `exwm-manage--kill-buffer-query-function`)

numerous times in the last couple of days.

Looking at the issue tracker, I see that this issue keeps re-appearing now and then.  And EXWM developers have been addressing this issue on a case-by-case basis.  

Assuming that I can reproduce the this or another *recurrent* bug, I can

1. Note down the timestamp
2. Reproduce the bug.
3. Copy the logs since timestamp noted in (1) to now.
4. Send the above log to EXWM issue tracker.

This way you see only the log relevant to issue at hand and not anything else.  (There is always some X event happening, the log keeps on growing and one can be overwhelmed or clueless about where to look)

As I type this, I see an option to /mark/ a location in the debug buffer.  

I believe this /mark/ functionality serves the same purpose as the one I am suggesting.  (The timestamp option has the advantage that it is quite intuitive what one has to do.  The option to put a /mark/ in the debug buffer is geeky and demands that user look at the source code or documentation)


 
> Please, try the following in `*scratch*`:
> 
> ```emacs-lisp
> (defun exwm-debug--hide-header-line (button)
>   (setq-local header-line-format nil)
>   (force-mode-line-update))
> 
> (setq-local header-line-format 
>             (with-temp-buffer
>               (insert "Messages appear in chronological order. ")
>               (insert-text-button "Hide hint"
>                                   'face custom-button
>                                   'follow-link t
>                                   'action 'exwm-debug--hide-header-line)
>               (buffer-string)))
> ```
> 
> Could you suggest a better informational message?

A fancy button is not necessary.  A simple hint would do. 

When I evalled your suggestion, I ended up with a button in my *scratch* buffer. I modified your  snippet so that it adds the button to the debug buffer.

```
(when exwm-debug
  (with-current-buffer (get-buffer xcb-debug:buffer)
    (setq-local header-line-format
		(with-temp-buffer
		  (insert-text-button "Read the log in chronological order"
				      'face custom-button
				      'follow-link t
				      'help-echo "Hide this hint"
				      'action (defun exwm-debug--hide-header-line (button)
						(with-current-buffer (get-buffer xcb-debug:buffer)
						  (setq-local header-line-format nil)
						  (force-mode-line-update))))
		  (buffer-string)))
    (force-mode-line-update)))
```

**Comment by ch11ng** [comment]
> I really didn't know if the logs are in chronological order or reverse chronological order.

Just like `*Messages*` most logs are implicitly chronological.  Of course there are a few exceptions like `git log`.

Timestamps can be useful when examining user-uploaded logs.  Perhaps `(emacs-uptime "%s")` is more suitable if users considered absolute time sensitive.  Does it make sense to you @medranocalvo ?

**Comment by medranocalvo** [comment]
@ch11ng I think that's great.

Nonetheless, I wouldn't discard including a hint indicating users how to use the log.  It could mention `xcb-debug:clear`, for example (@emacksnotes: that command clears the buffer), or `xcb-debug:mark`; it could even offer them as buttons.  Alternatively, it could refer to the Wiki for debugging instructions (https://github.com/ch11ng/exwm/wiki/EXWM-User-Guide#how-to-report-a-bug).

**Comment by ch11ng** [comment]
> Nonetheless, I wouldn't discard including a hint indicating users how to use the log.

Nice! Speaking of this, we forgot to make a menu entry for the debugging feature.  It should be included as well.

**Comment by emacksnotes** [comment]
> Add timestamps to `*XELB-DEBUG*`

Here is a demonstration of a sample debug session with the above feature: https://github.com/ch11ng/exwm/issues/628#issuecomment-525373170

Note that **I not only log the time, but also add a handwritten note to the *XELB-DEBUG*` file.**

**Comment by ch11ng** [comment]
@emacksnotes I'll add timestamps but not notes.  Writing notes itself generates some logs (mnibuffer setup/exit, keystrokes, depending on your config), which can be distracting.

**Comment by ch11ng** [comment]
Added in 2d36241a10b4b705e1482322383f135bca7ac184 and ch11ng/xelb@599ef709072c18022f5b08c0367721d126b1a716 .

**Comment by emacksnotes** [comment]
> https://github.com/ch11ng/exwm/commit/2d36241a10b4b705e1482322383f135bca7ac184#diff-b2301b53c4e2272d63fb6047c851c925R43

I was trying to collect debug logs for a iexwm issue today, and to change the default timestamp format.   So, I did

M-x customize-group RET exwm-debug

and was surprised to find that I couldn't find the newly added option. 

(I think) you should put `exwm-debug-log-time-function` in `exwm-debug`.  Right now it is under `exwm` group.

**Comment by emacksnotes** [comment]
> I was trying to collect debug logs for a exwm issue today, and wanted absolute timestamps in the logs. 

With `Exwm Debug Log Time Function` set to `exwm-debug-log-time`,  I was getting debug logs like this

```
[15:16:27] exwm--set-geometry:	Setting #x3000027 to 1366x749+0+18
[02:56:32] xcb:-cache-request:	Cache request #136648: [12 0 7 0 39 0 0 3 15 0 0 0 0 0 0 0 18 0 0 0 86 5 0 0 237 2 0 0]
[15:16:27] exwm--set-geometry:	Setting #x1e00143 to 1366x749+nil+nil
[02:56:32] xcb:-cache-request:	Cache request #136649: [12 0 5 0 67 1 224 1 12 0 0 0 86 5 0 0 237 2 0 0]
```

Note the mix of timestamps. 

I was a bit confused if XCB and EXWM are logging the same time, but in different formats i.e., I thought one was using a 12-hr format while the other was using 24-hr format.  

Then I realized that there are TWO DIFFERENT logging functions 

1. `exwm--log` (which calls `xcb-debug:message`) 
2. `xcb:-log` 

The two logign functions have THEIR OWN  DIFFERENT time formats

1. `exwm-debug-log-time-function` 
2. xcb-debug:log-time-function`

BUT they MULTIPLEX on to the same debug file `xcb-debug:buffer`.

Having two different formats in THE SAME debug log is confusing.  Either 

1.  Make `xcb` and `exwm` use two different debug files
2.  Or if they are using the same debug file, they should use the same timestamp format.

Did you _deliberately_ chosen the log format to be this way .... 

_(I have re-opened this bug)_

**Comment by ch11ng** [comment]
> Would you consider changing the `:group` option of `Exwm Debug Log Time Function` to `exwm-debug`?

Sure.  Didn't notice there was an `exwm-debug` section.  It's automatically generated by `define-minor-mode`.

> Make xcb and exwm use two different debug files

This would make it hard to tell the order of logs come from different buffers.

> Or if they are using the same debug file, have them use the same timestamp format.

I'd go this way.

**Comment by ch11ng** [comment]
Fixed with 16d643dd3b2b8e468dbbfb2ad03d7abe41fdf94d.
